### PR TITLE
clang support

### PR DESCRIPTION
**Pull request by dryganets**
The code compiles with clang.

Test Plan:
----------
If it builds - it works. There is no semantic changes, I just fixed the warnings reported by clang.

Release Notes:
----------
[ANDROID] [Internal] [BuildSystem]

### CODE REVIEW

File: ReactAndroid/src/main/jni/Application.mk [hunk 1]
```diff
@@ -1,7 +1,7 @@
 APP_BUILD_SCRIPT := Android.mk
 
 APP_ABI := armeabi-v7a x86
-APP_PLATFORM := android-9
+APP_PLATFORM := android-14
```
Comment (dryganets) [inline ReactAndroid/src/main/jni/Application.mk]: Fixes unified headers warning. You don't support android-9 anyway.

File: ReactAndroid/src/main/jni/first-party/yogajni/Android.mk [hunk 2]
```diff
@@ -10,7 +10,7 @@ LOCAL_SRC_FILES := \
 LOCAL_C_INCLUDES := $(LOCAL_PATH)/jni
 
 LOCAL_CFLAGS += -Wall -Werror -fvisibility=hidden -fexceptions -frtti -O3
-CXX11_FLAGS := -std=c++11
+CXX11_FLAGS := -std=c++14
```
Comment (dryganets) [inline ReactAndroid/src/main/jni/first-party/yogajni/Android.mk]: in clang lambdas you are using is part of c++14 extension.
the code continues to compile with your toolchain.

**Comment by react-linter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :clipboard: Release Notes - <i>This PR may have incorrectly formatted Release Notes.</i>
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ReactAndroid/src/main/jni/react/jni/AndroidJSCFactory.h [hunk 3]
```diff
@@ -10,7 +10,7 @@
 
 namespace folly {
 
-class dynamic;
+struct dynamic;
```
Comment (dryganets) [inline ReactAndroid/src/main/jni/react/jni/AndroidJSCFactory.h]: This is a struct not class see folly sources.
clang is more diligent about such warnings.

File: ReactAndroid/src/main/jni/Application.mk [hunk 4]
```diff
@@ -12,4 +12,4 @@ APP_STL := gnustl_shared
 # Make sure every shared lib includes a .note.gnu.build-id header
 APP_LDFLAGS := -Wl,--build-id
 
-NDK_TOOLCHAIN_VERSION := 4.8
+NDK_TOOLCHAIN_VERSION := 4.9
```
Comment (dryganets) [inline ReactAndroid/src/main/jni/Application.mk]: This is the best common denominator between clang and gcc.

File: ReactAndroid/src/main/jni/third-party/glog/Android.mk [hunk 5]
```diff
@@ -3,24 +3,24 @@ LOCAL_PATH:= $(call my-dir)
...
+LOCAL_C_INCLUDES += $(LOCAL_PATH) $(LOCAL_PATH)/.. $(LOCAL_PATH)/glog-0.3.5/src/
 
-LOCAL_EXPORT_C_INCLUDES := $(LOCAL_PATH)/.. $(LOCAL_PATH)/glog-0.3.3/src/
+LOCAL_EXPORT_C_INCLUDES := $(LOCAL_PATH)/.. $(LOCAL_PATH)/glog-0.3.5/src/
 
 LOCAL_CFLAGS += \
   -Wall \
   -Wwrite-strings \
   -Woverloaded-virtual \
-  -Wno-sign-compare \
```
Comment (dryganets) [inline ReactAndroid/src/main/jni/third-party/glog/Android.mk]: You shouldn't add this suppress.
The lib had a bug:
https://github.com/google/glog/commit/a5ffa884137f7687d0393ccba22557d583654a25

**Comment by hramos** [comment]
Thanks for the PR @dryganets! we've been making some changes to our Gradle setup, can you take a look at the conflicts?

**Comment by dryganets** [comment]
I'm on a week long vacation. Will rebase the branch as soon as get a grip on my laptop.
Thanks!

**Comment by dryganets** [comment]
All conflicts are resolved.
I synced the iOS and Android third-party-dependencies versions as well.


The patch to finally switch to clang looks like this:

```
From 9794123d1cf4bcc4b6d5244565cc2fab472b6834 Mon Sep 17 00:00:00 2001
From: Sergei Dryganets
Date: Mon, 9 Jul 2018 12:39:40 -0700
Subject: [PATCH] switched to clang

---
 ReactAndroid/src/main/jni/Application.mk               | 4 ++--
 ReactAndroid/src/main/jni/third-party/folly/Android.mk | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/ReactAndroid/src/main/jni/Application.mk b/ReactAndroid/src/main/jni/Application.mk
index 56318158cb..6d92747ec1 100644
--- a/ReactAndroid/src/main/jni/Application.mk
+++ b/ReactAndroid/src/main/jni/Application.mk
@@ -7,9 +7,9 @@ APP_MK_DIR := $(dir $(lastword $(MAKEFILE_LIST)))

 NDK_MODULE_PATH := $(APP_MK_DIR)$(HOST_DIRSEP)$(THIRD_PARTY_NDK_DIR)$(HOST_DIRSEP)$(REACT_COMMON_DIR)$(HOST_DIRSEP)$(APP_MK_DIR)first-party

-APP_STL := gnustl_shared
+APP_STL := c++_shared

 # Make sure every shared lib includes a .note.gnu.build-id header
 APP_LDFLAGS := -Wl,--build-id

-NDK_TOOLCHAIN_VERSION := 4.9
+NDK_TOOLCHAIN_VERSION := clang
diff --git a/ReactAndroid/src/main/jni/third-party/folly/Android.mk b/ReactAndroid/src/main/jni/third-party/folly/Android.mk
index e414966c6b..b0e7ec3797 100644
--- a/ReactAndroid/src/main/jni/third-party/folly/Android.mk
+++ b/ReactAndroid/src/main/jni/third-party/folly/Android.mk
@@ -19,7 +19,7 @@ LOCAL_CFLAGS += -Wall -Werror
 CXX11_FLAGS := -std=c++14
 LOCAL_CFLAGS += $(CXX11_FLAGS)

-FOLLY_FLAGS := -DFOLLY_NO_CONFIG=1 -DFOLLY_HAVE_CLOCK_GETTIME=1
+FOLLY_FLAGS := -DFOLLY_NO_CONFIG=1 -DFOLLY_HAVE_CLOCK_GETTIME=1 -DFOLLY_USE_LIBCPP=1 -DFOLLY_HAVE_MEMRCHR=1
 LOCAL_CFLAGS += $(FOLLY_FLAGS)

 LOCAL_EXPORT_CPPFLAGS := $(CXX11_FLAGS) $(FOLLY_FLAGS)
--
2.18.0
```

**This patch will work only starting from android-ndk-r16b, before r16b c++stl is incorrectly configured, so some throw_* defenitions won't be found.**

**Comment by dryganets** [comment]
The ObjC tests failure looks like some CI issue:
echo 'Packager did not respond. Retrying for attempt number 33

**Comment by gengjiawen** [comment]
@hramos Can you take a look this pr, I think this can help us with the folly upgrade.

**Comment by gengjiawen** [comment]
@dryganets Does this pr allow us build with ndk 16 or 17 ?

**Comment by dryganets** [comment]
@gengjiawen, the build with 16 and 17 should already work see this PR:
https://github.com/facebook/react-native/pull/18814

This PR focuses on the issues specific to the clang compiler.

You can't build with clang without upgrading to at least NDK-16b because of the incorrect libc++stl headers coming with ndk. Also you need to configure folly a little bit different.

I was able to build react-native with ndk-15c by patching the folly locally, but I don't recommend doing it - it is better to upgrade to the newer ndk version.

Once this change will be merged - it will be possible to move to clang with the simple patch I posted in this PR. It is as easy as switching the compiler and updating the folly flags.

**Comment by gengjiawen** [comment]
From what I test, looks like still need more effort with r17b
```
In file included from ./JSCPerfLogging.cpp:8:
In file included from D:/Developer/react-native/ReactAndroid/../ReactCommon/cxxreact/..\jschelpers/JSCHelpers.h:8:
In file included from d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\algorithm:62:
In file included from d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\bits/stl_algo.h:59:
d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\cstdlib:72:10: fatal error: 'stdlib.h' file not found
#include <stdlib.h>
         ^~~~~~~~~~
In file included from ./AndroidJSCFactory.cpp:6:
In file included from D:/Developer/react-native/ReactAndroid/../ReactCommon/cxxreact/..\cxxreact/JSCExecutor.h:9:
In file included from d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\memory:72:
In file included from d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\iosfwd:40:
In file included from d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\bits/postypes.h:40:
d:/android-sdk-windows/ndk-bundle/build//../sources/cxx-stl/gnu-libstdc++/4.9/include\cwchar:44:10: fatal error: 'wchar.h' file not found
#include <wchar.h>

```

**Comment by dulmandakh** [comment]
NDK r17b landed https://github.com/facebook/react-native/commit/6117a6c7205c969f93d39ba02e0583881572d5fa, and closing this issue.

**Pull request closed without merging**